dist_example.py

```diff
@@ -151,66 +91,93 @@ def test(args, model, device, test_loader, verbose=True):
+        new_model.eval()
+        model_future = client.scatter(new_model, broadcast=True)
+        # compute grads
+        grad_start = time.time()
+        grads = _get_gradients(client, model_future, n=len(train_set), idx=model_updates, batch_size=bs, n_workers=n_workers)  # a call to Dask
+        grad_time = time.time() - grad_start
+        # update SGD
+        opt.zero_grad()
+        num_data = sum(info["_num_data"] for info in grads)
+        assert num_data == bs
+        for name, param in model.named_parameters():
+            grad = sum(grad[name] for grad in grads)
+            # --- move the averaging to get_gradiations
+            param.grad = grad / num_data
 
-if __name__ == "__main__":
-    # Training settings
-    args = SimpleNamespace(
-        batch_size=64,
-        test_batch_size=1000,
-        epochs=14,
-        gamma=0.7,
-        lr=1.0,
-        no_cuda=False,
-        save_model=False,
-        seed=1,
-    )
-    use_cuda = not args.no_cuda and torch.cuda.is_available()
-
-    torch.manual_seed(args.seed)
-
-    device = torch.device("cuda" if use_cuda else "cpu")
-
-    kwargs = {"num_workers": 1, "pin_memory": True} if use_cuda else {}
-
-    transform = transforms.Compose(
-        [transforms.ToTensor(), transforms.Normalize((0.1307,), (0.3081,))]
-    )
-    train_set = datasets.MNIST(
-        "../data", train=True, download=True, transform=transform,
-    )
-    # For quicker debugging uncomment this line
-    #  train_set, _ = torch.utils.data.random_split(train_set, [2000, len(train_set) - 2000])
-
-    _train_loader = torch.utils.data.DataLoader(
-        train_set, batch_size=10_000, shuffle=False, **kwargs
-    )
-
-    # _train_loader's shuffle=False is importnat for these lines, otherwise
-    # we're looping twice over the data, and it gets shuffled in- the meantime
-    inputs = torch.cat([input.clone() for input, _ in _train_loader])
-    targets = torch.cat([target.clone() for _, target in _train_loader])
+        # update metrics!
+        metrics.append({ 'grad_time': grad_time, 'bs': bs, 'n_workers': n_workers, 'loop_time': time.time() - loop_start, 'kwargs': kwargs, 'idx': model_updates })
 
-    test_set = datasets.MNIST("../data", train=False, transform=transform)
-    test_loader = torch.utils.data.DataLoader(
-        test_set, batch_size=args.test_batch_size, shuffle=True, **kwargs
-    )
+        if model_updates % kwargs["save_freq"] == 0:
+            print("Saving metrics to disk")
+            pd.DataFrame(metrics).to_csv("loop_metrics" + log_suffix + ".csv")
 
-    model = Net().to(device)
-    optimizer = optim.Adadelta(model.parameters(), lr=args.lr)
+    # have last entry have overall data: total time, client init time, send to data to clinet time
+    overview_metrics = pd.DataFrame.from_dict({ 'train_time': time.time() - start_time, 'client_init': client_init })
+    loop_metrics = pd.DataFrame(metrics)
+    overview_metrics.to_csv("overview_metrics" + log_suffix + "csv")
```

Comment:
I wouldn't expect this function to have any side effects like writing to disk. I'd return the two DataFrames, then write them in the `if __name__ == "__main__"` block.

Reply:
Also, typo?

``` diff
- log_suffix + "csv"
+ log_suffix + ".csv"
```

Reply:
How can I do this while still saving metrics every N metrics?

Reply:
Good point. Writing to disk provides utility. Let's keep it in, and add a note in the documentation.

> How can I do this while still saving metrics to disk every N metrics?

In the past, I've found it use classes instead of functions. This allows easier use; I can interrupt the program part way through and recover partial results. 

Here's what I'd probably do for this class:

<details>

``` python
from sklearn.base import BaseEstimator

class Trainer(BaseEstimator):
    def __init__(self, fname):
        self.fname = fname
        super().__init__()

    def save(self):
        with open(self.fname, "w) as f:
            pickle.dump(self, f)

    def partial_fit(self):
        ...  # one gradient calculation of for-loop

if __name__ == "__main__":
    run = Trainer(fname="foo.pkl")
    for k in range(10):
        if k % 10 == 0:
            run.save()
        run.partial_fit()
```

</details>

To be clear, this is only a comment without any urgency.